### PR TITLE
chore: prisma -> drizzle

### PR DESCRIPTION
**Pull request by JoltCode**
# Description

Remove Prisma from deps, convert all code from using Prisma to using Drizzle

Fixes N/A (issue)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Tested through preexisting tests (throughout the codebase)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works 
- [x] I have updated .env.example if I added a new environment variable
- [x] My PR title follows the [semantic commits](https://www.conventionalcommits.org/en/v1.0.0/) style

### CODE REVIEW

File: apps/discord-bot/package.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
 		"@answeroverflow/db-mock": "workspace:^",
 		"@answeroverflow/discordjs-react": "^0.0.36",
 		"@answeroverflow/discordjs-utils": "workspace:^",
+		"@answeroverflow/elastic-types": "workspace:^",
```
Comment (RhysSullivan) [inline apps/discord-bot/package.json:16]: why did we add elastic types to discord bot

File: apps/discord-bot/src/listeners/parity/cache-parity.test.ts [hunk 2]
```diff
@@ -31,9 +29,7 @@ beforeEach(async () => {
 	client = await setupAnswerOverflowBot();
 	guild = mockGuild(client);
 	member = mockGuildMember({ client, guild });
-	user = await prisma.user.create({
-		data: {},
-	});
+	user = await db.insert(users).values({}); // TODO: set values (prisma -> drizzle)
```
Comment (RhysSullivan) [inline apps/discord-bot/src/listeners/parity/cache-parity.test.ts:32]: we should maybe make this not a leaky abstraction and just make a func in the  db package

File: apps/discord-bot/src/listeners/parity/cache-parity.test.ts [hunk 3]
```diff
@@ -6,12 +6,7 @@ import {
 	mockGuildMember,
 } from '@answeroverflow/discordjs-mock';
 import { setupAnswerOverflowBot } from '~discord-bot/test/sapphire-mock';
-import {
-	createDiscordAccount,
-	prisma,
-	type User,
-	_NOT_PROD_createOauthAccountEntry,
-} from '@answeroverflow/db';
+import { db } from '@answeroverflow/db';
```
Comment (RhysSullivan) [inline apps/discord-bot/src/listeners/parity/cache-parity.test.ts:9]: lets not use the drizzle db outside of the db package

File: apps/discord-bot/src/utils/conversions.ts [hunk 4]
```diff
@@ -11,14 +11,14 @@ import {
...
-	getDefaultServer,
-	type Message as AOMessage,
-	type Channel as AOChannel,
 	type Server as AOServer,
+	type Channel as AOChannel,
 	type DiscordAccount as AODiscordAccount,
-	getDefaultChannelWithFlags,
-} from '@answeroverflow/db';
+} from '@answeroverflow/db/src/schema';
+import { type Message as AOMessage } from '@answeroverflow/elastic-types';
```
Comment (RhysSullivan) [inline apps/discord-bot/src/utils/conversions.ts:18]: lets export this type from the db pacakge

File: apps/discord-bot/test/setup.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import { cleanupRedis } from '@answeroverflow/cache';
-import { prisma } from '@answeroverflow/db';
+import { db } from '@answeroverflow/db';
```
Comment (RhysSullivan) [inline apps/discord-bot/test/setup.ts:2]: move into a function in the db pacakge

File: apps/main-site/package.json [hunk 6]
```diff
@@ -41,6 +41,7 @@
 		"@trpc/react-query": "^10.37.1",
 		"@trpc/server": "^10.37.1",
 		"core-js": "^3.32.1",
+		"drizzle-orm": "^0.28.5",
```
Comment (RhysSullivan) [inline apps/main-site/package.json:44]: dont have drizzle in main site

File: packages/db/drizzle/schema.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+				table.inviteCode,
+			),
+		};
+	},
+);
+
+export const discordAccount = mysqlTable(
+	'DiscordAccount',
+	{
+		id: varchar('id', { length: 191 }).notNull(),
```
Comment (RhysSullivan) [inline packages/db/drizzle/schema.ts:74]: unqiue

File: packages/db/index.ts
Comment (RhysSullivan) [inline packages/db/index.ts:1]: removing all the barrel exports will probably break things, may be worth doing in a future PR instead to keep this one manageable

**Review by RhysSullivan — CHANGES_REQUESTED** [review]

File: apps/discord-bot/package.json (hunk 1 above)
Comment (JoltCode) [reply apps/discord-bot/package.json:16]: make db export these types - import directly from db now

File: packages/auth/src/adapter.test.ts [hunk 8]
```diff
@@ -1,8 +1,10 @@
 import { extendedAdapter } from './adapter';
 import type { z } from 'zod';
 import type { zUserSchema } from '@answeroverflow/cache';
-import { prisma } from '@answeroverflow/db';
+import { db } from '@answeroverflow/db';
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/auth/src/adapter.test.ts:4]: agreed to use db in auth package now

File: packages/auth/package.json [hunk 9]
```diff
@@ -22,9 +22,9 @@
 		"@answeroverflow/db": "workspace:^",
 		"@answeroverflow/env": "workspace:^",
 		"@answeroverflow/utils": "workspace:^",
-		"@next-auth/prisma-adapter": "^1.0.7",
-		"@prisma/client": "^4.16.2",
+		"@auth/drizzle-adapter": "^0.3.2",
 		"dotenv-cra": "^3.0.3",
+		"drizzle-orm": "^0.28.5",
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/auth/package.json:27]: agreed to use db in auth package now

File: packages/auth/src/adapter.test.ts [hunk 10]
```diff
@@ -52,12 +52,10 @@ describe('Discord Auth', () => {
 	});
 	// We have first seen their account on Discord from indexing their messages, we are linking their indexed account to what was signed in with
 	it('should link to a Discord user for an existing account', async () => {
-		await prisma.discordAccount.create({
-			data: {
-				id: mockDiscordAccount.id,
-				name: mockDiscordAccount.username,
-				avatar: mockDiscordAccount.avatar,
-			},
+		await db.insert(discordAccounts).values({
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/auth/src/adapter.test.ts:55]: agreed to use db in auth package now

File: packages/cache/src/discord.ts [hunk 11]
```diff
@@ -24,11 +23,11 @@ export async function discordApiFetch(
 		},
 	});
 	if (data.status === 401) {
-		const account = await prisma.account.findFirst({
-			where: {
-				provider: 'discord',
-				access_token: callOpts.accessToken,
-			},
+		const account = await db.query.accounts.findFirst({
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/cache/src/discord.ts:26]: agreed to use db in auth package now

File: packages/auth/src/adapter.ts [hunk 12]
```diff
@@ -1,10 +1,13 @@
-import { prisma, upsertDiscordAccount } from '@answeroverflow/db';
-import { PrismaAdapter } from '@next-auth/prisma-adapter';
+import { db } from '@answeroverflow/db';
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/auth/src/adapter.ts:1]: agreed to use db in auth package now

File: packages/db/drizzle.config.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+dotenv.config({
+	path: '../../.env',
+});
+
+export default {
+	schema: './schema.ts',
+	out: './drizzle',
+	driver: 'mysql2',
+	dbCredentials: {
+		connectionString: process.env['DATABASE_URL'] ?? '',
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/db/drizzle.config.ts:13]: missing env vars - something to look at in the future

File: packages/db/drizzle/0000_deep_miracleman.sql
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/db/drizzle/0000_deep_miracleman.sql:1]: 👍

File: packages/db/index.ts
Comment (JoltCode) [reply packages/db/index.ts:1]: 😭

Comment (JoltCode) [reply packages/db/index.ts:1]: will add to index.ts - we can slowly move them over in the future

**Review by RhysSullivan — APPROVED** [review]
Needs an actual review but tests are passing

File: packages/db/src/zodSchemas/discordAccountSchemas.ts [hunk 14]
```diff
@@ -1,6 +1,6 @@
-import type { DiscordAccount } from '@prisma/client';
 import { omit, pick } from '@answeroverflow/utils';
 import { z } from 'zod';
+import { DiscordAccount } from '../schema';
 
 type DiscordAccountZodFormat = {
 	[K in keyof DiscordAccount]: z.ZodTypeAny;
```
Comment (RhysSullivan) [inline packages/db/src/zodSchemas/discordAccountSchemas.ts:6]: Make this file like the others

File: packages/db/src/server.ts [hunk 15]
```diff
@@ -81,67 +84,68 @@ export async function updateServer({
...
+	const found = await db.query.servers.findMany({
+		where: inArray(servers.id, ids),
+	});
 	return found.map(addFlagsToServer);
 }
 
-export function upsertServer(input: z.infer<typeof zServerUpsert>) {
+export function upsertServer(
+	input: z.infer<typeof zServerUpsert> & {
+		create: z.infer<typeof zServerCreate> & {
```
Comment (RhysSullivan) [inline packages/db/src/server.ts:143]: This doesn't look right

File: packages/db/src/auth.ts [hunk 16]
```diff
@@ -94,33 +66,72 @@ export async function updateProviderAuthToken(
...
-	return prisma.account.create({
-		data: {
-			provider: 'discord',
-			type: 'oauth',
-			providerAccountId: discordUserId,
-			userId: userId,
-			access_token: getRandomId(),
-		},
+	await db.insert(accounts).values({
+		id: getRandomId(), // TODO: Is this okay?
```
Comment (RhysSullivan) [inline packages/db/src/auth.ts:108]: Remove todo

File: packages/db/src/db.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
+import { sharedEnvs } from '@answeroverflow/env/shared';
+
+import {
+	drizzle as psDrizzle,
+	PlanetScaleDatabase,
+} from 'drizzle-orm/planetscale-serverless';
+import { connect } from '@planetscale/database';
+import { fetch } from 'undici';
```
Comment (RhysSullivan) [inline packages/db/src/db.ts:8]: Do we need undici here? Look into

File: packages/db/src/db.ts [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+import { drizzle as mysqlDrizzle } from 'drizzle-orm/mysql2';
+import mysql from 'mysql2/promise';
+
+import * as schema from './schema';
+const connection = connect({
```
Comment (RhysSullivan) [inline packages/db/src/db.ts:14]: Move connection into if statement

File: packages/db/src/db.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+	url: sharedEnvs.DATABASE_URL,
+	fetch,
+});
+
+export const db: PlanetScaleDatabase<typeof schema> =
```
Comment (RhysSullivan) [inline packages/db/src/db.ts:19]: Add a comment about why we're doing this like this, it's so we can connect to non planetscale databases

File: packages/constants/src/analytics.ts [hunk 20]
```diff
@@ -1,10 +1,10 @@
-import type {
-	Channel,
-	ChannelWithFlags,
-	Server,
-	ServerWithFlags,
-} from '@answeroverflow/prisma-types';
 import type { Message } from '@answeroverflow/elastic-types';
+// eslint-disable-next-line no-restricted-imports
+import type { Channel, Server } from '../../db/src/schema';
```
Comment (RhysSullivan) [inline packages/constants/src/analytics.ts:3]: This is bad why are we importing like this

File: packages/auth/src/adapter.ts [hunk 21]
```diff
@@ -1,42 +1,207 @@
...
+import { accounts, db, users, verificationTokens } from '@answeroverflow/db';
 import { getDiscordUser } from '@answeroverflow/cache';
-import type { Adapter, AdapterAccount } from 'next-auth/adapters';
+import { sessions, tenantSessions } from '@answeroverflow/db/src/schema';
+import { upsertDiscordAccount } from '@answeroverflow/db/src/discord-account';
+import { and, eq } from 'drizzle-orm';
+import type { Adapter } from '@auth/core/adapters';
+import { randomUUID } from 'node:crypto';
 
 export const extendedAdapter: Adapter = {
```
Comment (RhysSullivan) [inline packages/auth/src/adapter.ts:9]: Add comment about us needing to make our own adapter due to drizzle not matching prisma adapter

File: packages/api/src/utils/operations.ts [hunk 22]
```diff
@@ -1,29 +1,10 @@
...
-export function addDefaultValues<T, F>(
-	input: T[],
-	getDefaultValue: (input: T) => F,
-) {
-	return input.map((v) => getDefaultValue(v));
-}
-
 export async function findOrThrowNotFound<T>(
-	find: () => Promise<T | null>,
+	find: () => Promise<T | null | undefined>,
```
Comment (RhysSullivan) [inline packages/api/src/utils/operations.ts:4]: We should look to moving to only null or only undefined for the code base in the future

File: apps/main-site/src/pages/api/auth/[...nextauth].ts [hunk 23]
```diff
@@ -17,6 +17,7 @@ export default async function handler(
 	if (token) {
 		const nextAuthSession = await findTenantSessionByToken(token);
 		// add a cookie to the request using the next auth header
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
Comment (RhysSullivan) [inline apps/main-site/src/pages/api/auth/[...nextauth].ts:20]: It's weird we had to add this line when nothing changed

File: apps/main-site/package.json [hunk 24]
```diff
@@ -5,7 +5,7 @@
 	"scripts": {
 		"with-env": "dotenv -e ../../.env --",
 		"dev": "pnpm with-env next dev",
-		"build": "(cd ../../ && pnpm db:generate && cd apps/main-site) && pnpm with-env next build",
```
Comment (RhysSullivan) [inline apps/main-site/package.json:8]: Remove cds

**Review by RhysSullivan — APPROVED** [review]

**Pull request merged**